rtl_tests.c

```diff
@@ -211,8 +211,80 @@ void test_RtlAssert(){
     /* FIXME: This is a stub! implement this function! */
 }
 
+// FIXME: RtlCaptureContext also sets Ebp, Eip, and Esp in the context record.
+// However, without doing some math there isn't a good way to know what Ebp, Esp, and Eip were
+// set to in the caller of RtlCaptureContext. And I'm too lazy to figure out that code right now.
 void test_RtlCaptureContext(){
-    /* FIXME: This is a stub! implement this function! */
+    const char* func_num = "0x0109";
+    const char* func_name = "RtlCaptureContext";
+    BOOL tests_passed = 1;
+    print_test_header(func_num, func_name);
+
+    CONTEXT result_context;
+    // cs and ss words so make sure that the upper word of the Seg* dword is 0 or the tests will fail
+    // due to unitialized values.
+    result_context.SegCs = 0;
+    result_context.SegSs = 0;
+
+    DWORD eax_val = 0;
+    DWORD ebx_val = 0x11223344;
+    DWORD ecx_val = 0x55667788;
+    DWORD edx_val = 0x99AABBCC;
+    DWORD esi_val = 0xDDEEFF11;
+    DWORD edi_val = 0xDEADBEEF;
+    DWORD flags_val = 0;
+    // Again, cs and ss are words so make sure the upper word is initialized to 0
+    DWORD cs_val = 0;
+    DWORD ss_val = 0;
+
+    // Make sure we can clobber all registers
+    asm("pusha\n");
+
+    // Push &result_context to pass into RtlCaptureContext
+    asm("leal %0, %%eax\n" : : "m" (result_context));
+    asm("pushl %eax\n");
+
+    // Store current values of cs and ss
+    asm("leal %0, %%eax\n" : : "m" (cs_val));
+    asm("movw %cs, (%eax)\n");
+    asm("leal %0, %%eax\n" : : "m" (ss_val));
+    asm("movw %ss, (%eax)\n");
+
+    // Store current value of the flags
+    asm("leal %0, %%eax\n" : : "m" (flags_val));
+    asm("pushf\n");
+    asm("popl (%eax)\n");
+
+    // Setup pre-set register values
+    asm("movl %0, %%ebx\n" : : "r" (ebx_val));
+    asm("movl %0, %%ecx\n" : : "r" (ecx_val));
+    asm("movl %0, %%edx\n" : : "r" (edx_val));
+    asm("movl %0, %%esi\n" : : "r" (esi_val));
+    asm("movl %0, %%edi\n" : : "r" (edi_val));
+
+    // Call RtlCaptureContext and cleanup registers
+    asm("movl __imp__RtlCaptureContext@4, %eax\n");
+    asm("movl %%eax, %0\n" : "=m" (eax_val) : );
+    asm("calll *%eax\n");
+    asm("popa\n");
```

Comment:
Shouldn't all of this code be in one large `asm` block, with the `volatile` keyword?

I'm not sure if the compiler might be allowed to add instructions between ASM blocks, but I'd assume so if it comes to clobbering.

Reply:
Good idea. I will rewrite the section to be one block of volatile asm code.